Update the Lineage guide for new clustering UX [do not merge until release] (#607)
wandb/docs: docs/guides/artifacts/explore-and-traverse-an-artifact-graph.md

    <title>Explore direct acyclic W&B Artifact graphs.</title>
</head>

> W&B automatically tracks the artifacts a given run logged as well as the artifacts a given run used. These artifacts can include datasets, models, evaluation results, or more. You can explore an artifact's lineage to track and manage of the various artifacts produced throughout the machine learning lifecycle.

Reviewer: I think there's a typo

Reviewer: artifact's lineage to track and manage of the various artifacts

Author: Fixed!

## Lineage
Tracking an artifact's lineage has several key benefits:

- Reproducibility: By tracking the lineage of all artifacts, teams can reproduce experiments, models, and results, which is essential for debugging, experimentation, and validating machine learning models.

> - Version Control: Artifact lineage involves versioning artifacts and tracking their changes over time. This allows teams to roll back to previous versions of data, code, or models if needed.

Reviewer: ```suggestion
- Version Control: Artifact lineage involves versioning artifacts and tracking their changes over time. This allows teams to roll back to previous versions of data or models if needed.
```

- Auditing: Having a detailed history of the artifacts and their transformations enables organizations to comply with regulatory and governance requirements.

- Collaboration and Knowledge Sharing: Artifact lineage facilitates better collaboration among team members by providing a clear record of attempts as well as what worked, and what didn’t. This helps in avoiding duplication of efforts and accelerates the development process.

### Finding an artifact's lineage
When selecting an artifact in the **Artifacts** tab, you can see your artifact's lineage. This graph view shows a general overview of your pipeline. 

To view an artifact graph:

1. Navigate to your project in the W&B App UI
2. Choose the artifact icon on the left panel.
3. Select **Lineage**.

![Getting to the Lineage tab](../../../static/images/artifacts/lineage1.gif)

### Navigating the lineage graph

> The artifact or run type you provide appears in front of its name, with artifacts represented by blue icons and runs represented by green icons. Arrows detail the input and output of a run or artifact on the graph. 

Reviewer: for runs - the type is called job type not run type

![Run and artifact nodes](../../../static/images/artifacts/lineage2.png)

:::info
You can view the type and the name of artifact in both the left sidebar and in the **Lineage** tab. 
:::

![Inputs and outputs](../../../static/images/artifacts/lineage2a.gif)

For a more detailed view, click any individual artifact or run to get more information on a particular object.

![Previewing a run](../../../static/images/artifacts/lineage3a.gif)

### Artifact clusters

When a level of the graph has five or more runs or artifacts, it creates a cluster. A cluster has a search bar to find specific versions of runs or artifacts and pulls an individual node from a cluster to continue investigating the lineage of a node inside a cluster. 

Clicking on a node opens a preview with an overview of the node. Clicking on the arrow extracts the individual run or artifact so you can examine the lineage of the extracted node.

![Searching a run cluster](../../../static/images/artifacts/lineage3b.gif)

## Use the API to track lineage
You can also navigate a graph using the [W&B API](../../ref/python/public-api/api.md). 

> Create an artifact object. First, create a run with `wandb.init`. Then,create a new artifact or retrieve an existing one with `wandb.Artifact`. Next, add files to the artifact with `.add_file`. Finally, log the artifact to the run with `.log_artifact`. The finished code looks something like this:

Reviewer: ```suggestion
Create an artifact. First, create a run with `wandb.init`. Then, create a new artifact or retrieve an existing one with `wandb.Artifact`. Next, add files to the artifact with `.add_file`. Finally, log the artifact to the run with `.log_artifact`. The finished code looks something like this:
```

    <title>Explore direct acyclic W&B Artifact graphs.</title>
</head>

> W&B automatically tracks the artifacts a given run logged as well as the artifacts a given run used. These artifacts can include data sets, models, parameters, metrics, and code. You can explore an artifact's lineage to track and manage of the various artifacts produced throughout the machine learning lifecycle.

Reviewer: ```suggestion
W&B automatically tracks the artifacts a given run logged as well as the artifacts a given run used. These artifacts can include datasets, models, eval results, or more. You can explore an artifact's lineage to track and manage of the various artifacts produced throughout the machine learning lifecycle.
```
I'm realizing we shouldn't tell ppl to store parameters, metrics, and code since those should be logged to runs but we can also just do models and datasets

## Lineage

> ### Benefits of tracking lineage

Reviewer 2: Let's remove this "### Benefits of tracking lineage" we can assume that the intro topic "##Lineage" describes why it's useful.

Tracking an artifact's lineage has several key benefits:

- Reproducibility: By tracking the lineage of all artifacts, teams can reproduce experiments, models, and results, which is essential for debugging, experimentation, and validating machine learning models.

- Version Control: Artifact lineage involves versioning of artifacts so that changes are tracked over time. This allows teams to roll back to previous versions of data, code, or models if needed.

- Auditing: Having a detailed history of the artifacts and their transformations enables organizations to comply with regulatory and governance requirements.

> - Collaboration and Knowledge Sharing: Artifact lineage facilitates better collaboration among team members by providing a clear record of what has been tried, what worked, and what didn’t. This helps in avoiding duplication of efforts and accelerates the development process.

Reviewer: ```suggestion
- Collaboration and Knowledge Sharing: Artifact lineage facilitates better collaboration by providing a clear record of which intermediate artifacts are produced. Intermediate artifacts can be re-used and pulled into downstream projects to avoid duplication.
```

### Finding an artifact's lineage
When selecting an artifact in the **Artifacts** tab, you can see your artifact's lineage. This graph view shows a general overview of your pipeline. 

To view an artifact graph:

1. Navigate to your project in the W&B App UI
2. Choose the artifact icon on the left panel.
3. Select **Lineage**.

> ![Getting to the Lineage tab](../../../static/images/artifacts/lineage1.gif)

Reviewer 2: Can you remake this .gif? It's too small to see (even when expanded).

### Navigating the lineage graph

> The lineage graph generates based on the `type` you provide when you create runs and artifacts. 

Reviewer: Re-reading this sentence - it doesn't actually make too much sense. We don't generate the graph based on type, we generate based off using log_artifact and use_artifact to mark inputs and outputs to runs. I say we remove or reframe how we generate (I think by generate this comment initially meant to point out out that the type of an artifact will be displayed on an artifacts node"

The artifact or run type you provide is located above its name, with artifacts represented by blue icons and runs represented by green icons. The input and output of a run or artifact is depicted in the graph with arrows. 

:::info
You can view the type and the name of artifact in both the left sidebar and in the **Lineage** tab. 
:::

![Run and artifact nodes](../../../static/images/artifacts/lineage2.png)

For a more detailed view, click the arrow on any individual artifact or run to get more information on a particular object.

### Artifact clusters

When a level of the graph has five or more runs or artifacts, it creates a cluster. A cluster has a search bar to find specific versions of runs or artifacts and pulls an individual node from a cluster to continue investigating the lineage of a node inside a cluster. 

![Searching a run cluster](../../../static/images/artifacts/lineage3.gif)

> Clicking on a node opens a preview with an overview of the node. Clicking on the arrow pulls out the run or artifact so you can track the lineage of the individual node.

Reviewer: ```suggestion
Clicking on a node opens a preview with an overview of the node. Clicking on the arrow extracts the individual run or artifact so you can examine the lineage of the extracted node. 
```

![Getting more detail from a node](../../../static/images/artifacts/lineage4.gif)

## Use the API to track lineage

> You can also navigate a graph usuing the [W&B API]((../../ref/python/public-api/api.md)). 

Reviewer: ```suggestion
You can also navigate a graph using the [W&B API]((../../ref/python/public-api/api.md)). 
```

> You can also navigate a graph usuing the [W&B API]((../../ref/python/public-api/api.md)). 

Reviewer 2: There's an extra ")" floating here.

Create an artifact object. First, create a run with `wandb.init`. Then,create a new artifact or retrieve an existing one with `wandb.Artifact`. Next, add files to the artifact with `.add_file`. Finally, log the artifact to the run with `.log_artifact`. The finished code looks something like this:

> ```

Reviewer 2: Specify the code type (python) https://docusaurus.io/docs/markdown-features/code-blocks#syntax-highlighting

> W&B automatically tracks the artifacts a given run logged as well as the artifacts a given run used. Explore the lineage of an artifact with the W&B App UI or programmatically.

Reviewer: Do you think we can add a blurb (maybe at the end) as to what the benefits are of having your project's lineage tracked? Something like:

Use an artifact's lineage to track and manage of the various artifacts produced throughout the machine learning lifecycle. These artifacts can include data sets, models, parameters, metrics, and code. Artifact lineage is crucial for ensuring reproducibility, auditability, and governance of machine learning projects.

The key components and benefits of artifact lineage in MLOps include:

Reproducibility: By tracking the lineage of all artifacts, teams can reproduce experiments, models, and results, which is essential for debugging, experimentation, and validating machine learning models.

Version Control: Artifact lineage involves versioning of artifacts so that changes can be tracked over time. This allows teams to roll back to previous versions of data, code, or models if needed.

Auditability: Having a detailed history of the artifacts and their transformations enables organizations to comply with regulatory and governance requirements. This is particularly important in industries such as finance and healthcare, where models must be explainable and transparent.

Collaboration and Knowledge Sharing: Artifact lineage facilitates better collaboration among team members by providing a clear record of what has been tried, what worked, and what didn’t. This helps in avoiding duplication of efforts and accelerates the development process.

> W&B automatically tracks the artifacts a given run logged as well as the artifacts a given run used. Explore the lineage of an artifact with the W&B App UI or programmatically.

Reviewer: Thinking we should start with a more generic section that covers "what you can expect to find in your artifact's lineage graph" or "what is in an artifact's lineage graph" or "anatomy of your artifact's lineage graph". My proposal:

An artifact's lineage will be displayed as Directed Acyclic Graph (DAG) that visualizes the relationships and dependencies between various artifacts within a W&B project. Artifacts in W&B can be datasets, models, or any other relevant assets that are part of your ML pipeline. The DAG provides a graphical representation of how these artifacts evolve, interact, and contribute to the development of machine learning models. When you view an artifact lineage DAG in W&B, you will see:

- Nodes: Each node in the DAG represents either an artifact version or run. An artifact version could be, a dataset or model version, a launch job, or a set of evaluation results. A run could be training run, or a pre-processing transformation. 

- Edges: Edges connect the nodes and represent the flow or the dependency between artifacts. 

- Versioning: The DAG highlights the version history of an artifact, allowing users to track changes over time. This helps identify the exact version of an artifact used or logged by a W&B run. 

- Provenance: Users can see the provenance of artifacts, meaning the origin and the transformation process that an artifact went through.

> W&B automatically tracks the artifacts a given run logged as well as the artifacts a given run used. Explore the lineage of an artifact with the W&B App UI or programmatically.

Reviewer: We should add gifs + section for cluster specific interactions coming in with new UX:

- Summary of Changes:
    - When a level of the graph (i.e a vertical column of nodes) has 5+ runs or artifacts, a cluster will be created
    - A cluster has a search bar to find specific versions
    - To continue investigating the lineage of a node in a cluster, a node can be pulled out from a cluster
    - Clicking on a node, whether an artifact or run node, will open a preview drawer with an overview of the node (previously a new tab opened, redirecting the user to a new lineage graph)

I don't see any details of these interactions in these PR. 

https://www.notion.so/wandbai/DAG-Lineage-Clustering-Performance-Roll-out-Plan-1a0afea7590f4b7bbd7ab26cd7dec3ed?pvs=4#a3b14cdde82d4580a400416e22e2e53c

## Lineage

When selecting an artifact in the **Artifacts** tab, you can see your artifact's lineage. This graph view shows a general overview of your pipeline. 

To view an artifact graph:

1. Navigate to your project in the W&B App UI
2. Choose the artifact icon on the left panel.
3. Select **Lineage**.

### Navigating lineage 
The lineage graph is generated based on the `type` you provide when you create runs and artifacts. 

> The artifact or run type you provide is located above its name, withA artifacts represented by blue icons and runs represented by green icons. The input and output of a run or artifact is depicted in the graph with arrows. 

Reviewer: ```suggestion
The artifact or run type you provide is located above its name, with artifacts represented by blue icons and runs represented by green icons. The input and output of a run or artifact is depicted in the graph with arrows. 
```

Reviewer: added a few more details here to incorporate in comment above that might be appropriate for this section - https://github.com/wandb/docodile/pull/607/files#r1515154412

:::info
You can view the type and the name of artifact in both the left sidebar and in the **Lineage** tab. 
:::

For a more detailed view, click the arrow on any individual artifact or run to get more information on a particular object.

### Artifact clusters

When a level of the graph has five or more runs or artifacts, a cluster is created. A cluster has a search bar to find specific versions of runs or artifact and an individual node can be pulled from a cluster to continue investigating the lineage of a node inside a cluster. Clicking on a node opens a preview with an overview of the node.

## Traverse a graph programmatically 
You can also navigate a graph usuing the [W&B API]((../../ref/python/public-api/api.md)). 

> ### Traverse from an artifact

Reviewer: Hmm I'm a bit confused by the wording here - what does it mean by traverse from an artifact?

> Create an artifact object and provide the name of the project, artifact, and alias of the artifact:

Reviewer: I think we can remove mentions of object since artifact and object are a bit redundant

> Create an artifact object and provide the name of the project, artifact, and alias of the artifact:

Reviewer: It might be worth to avoid using public API for artifact creation and instead show users how to create it in the context of a run. The workflow would be - create an artifact, add content to it, and log it to a run (this might be a training run, where the logged output is a model). And then then show how to then mark this artifact as input to a downstream run. 

Then continue with existing code of how to walk up and down from that newly created artifact. 

For example - [this](https://docs.wandb.ai/guides/artifacts/create-a-new-artifact-version#single-run)
<img width="696" alt="Screenshot 2024-03-06 at 2 21 20 PM" src="https://github.com/wandb/docodile/assets/40642416/0a9217ab-b4bc-45ad-9f83-9835f968ea29">
section shows artifact creation in context of a run.

consumer_runs = artifact.used_by()
```

### Traverse from a run

> Create an artifact object and provide the name of the project, artifact, and alias of the artifact:

Reviewer: ```suggestion
Create an artifact object and provide the name of the project, artifact, and version number or alias of the artifact:
```

producer_run = artifact.logged_by()
consumer_runs = artifact.used_by()
```

> ## Next steps

Reviewer: can we give users a public project to explore the lineage in product? Here is a good project that can be used - https://wandb.ai/wandb-smle/artifact_workflow/artifacts/raw_dataset/raw_data/v0/lineage
